File: pkg/cli/auth/aws/aws.go
package aws

import (
	"context"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sts"
	log "github.com/sirupsen/logrus"
	"goauthentik.io/platform/pkg/agent_local/client"
	"goauthentik.io/platform/pkg/pb"
)

type CredentialsOpts struct {
	Profile  string
	ClientID string
	// AWS specific things
	RoleARN string
	Region  string
}

type AWSCredentialOutput struct {
	Version         int
	AccessKeyId     string
	SecretAccessKey string
	SessionToken    string
	Expiration      time.Time
}

func (aco AWSCredentialOutput) Expiry() time.Time {
	return aco.Expiration
}

func GetCredentials(c *client.Client, ctx context.Context, opts CredentialsOpts) *AWSCredentialOutput {
	log := log.WithField("logger", "auth.aws")

	cc := client.NewCache[AWSCredentialOutput](c, &pb.RequestHeader{
		Profile: opts.Profile,
	}, "auth-aws-cache", opts.RoleARN)
	if v, err := cc.Get(ctx); err == nil {
		log.Debug("Got AWS Credentials from cache")
		return &v
	}

	stsc := sts.New(sts.Options{
		Region: opts.Region,
	})
	res, err := c.CachedTokenExchange(ctx, &pb.TokenExchangeRequest{
		Header: &pb.RequestHeader{
			Profile: opts.Profile,
		},
		ClientId: opts.ClientID,
	})
	if err != nil {
		log.WithError(err).Fatal("failed to exchange token")
		return nil
	}

	curr, err := c.GetCurrentToken(ctx, &pb.CurrentTokenRequest{
		Header: &pb.RequestHeader{
			Profile: opts.Profile,
		},
		Type: pb.CurrentTokenRequest_VERIFIED,
	})
	if err != nil {
		log.WithError(err).Fatal("failed to get current token")
		return nil
	}

	log.Debug("Fetching AWS Credentials...")
	a, err := stsc.AssumeRoleWithWebIdentity(ctx, &sts.AssumeRoleWithWebIdentityInput{
		RoleArn:          aws.String(opts.RoleARN),
		RoleSessionName:  aws.String(curr.Token.PreferredUsername),
		WebIdentityToken: aws.String(res.AccessToken),
	})
	if err != nil {
		log.WithError(err).Panic("failed to assume WebIdentity")
		return nil
	}
	output := AWSCredentialOutput{
		Version:         1,
		AccessKeyId:     *a.Credentials.AccessKeyId,
		SecretAccessKey: *a.Credentials.SecretAccessKey,
		SessionToken:    *a.Credentials.SessionToken,
		Expiration:      *a.Credentials.Expiration,
	}
	err = cc.Set(ctx, output)
	if err != nil {
		log.WithError(err).Warning("failed to cache AWS Credentials")
	}
	return &output
}

File: ak-cli/src/auth/aws.rs
use crate::cache::{CacheData, ClientCache};
use eyre::{bail, Result, WrapErr};
use ak_platform::{
    client::user::{AnyService, Client},
    generated::{
        agent::RequestHeader,
        agent_auth::{CurrentTokenRequest, TokenExchangeRequest, current_token_request},
    },
    grpc::assert_response_valid,
};
use aws_types::{SdkConfig, region::Region};
use pbjson_types::Timestamp;
use serde::{Deserialize, Serialize};

#[derive(Clone)]
pub struct CredentialsOpts {
    pub profile: String,
    pub client_id: String,
    // AWS specific things
    pub role_arn: String,
    pub region: String,
}

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "PascalCase")]
pub struct AWSCredentialOutput {
    version: i32,
    access_key_id: String,
    secret_access_key: String,
    session_token: String,
    expiration: Timestamp,
}

impl CacheData for AWSCredentialOutput {
    fn expiry(&self) -> Timestamp {
        self.expiration
    }
}

pub async fn get_credentials(
    c: Client<AnyService>,
    opts: CredentialsOpts,
) -> Result<AWSCredentialOutput> {
    let cc = ClientCache::new(
        c.clone(),
        RequestHeader {
            profile: opts.profile.clone(),
        },
        vec!["auth-aws-cache".to_string(), opts.role_arn.clone()],
    );
    if let Ok(v) = cc.get().await {
        return Ok(v);
    }

    let config = &SdkConfig::builder()
        .region(Region::new(opts.region.clone()))
        .build();
    let sts = aws_sdk_sts::Client::new(config);

    let res = c
        .clone()
        .auth()
        .cached_token_exchange(TokenExchangeRequest {
            header: Some(RequestHeader {
                profile: opts.profile.clone(),
            }),
            client_id: opts.client_id.clone(),
        })
        .await
        .wrap_err("failed to exchange token")?
        .into_inner();
    assert_response_valid(res.header)?;

    let curr = c
        .clone()
        .auth()
        .get_current_token(CurrentTokenRequest {
            header: Some(RequestHeader {
                profile: opts.profile.clone(),
            }),
            r#type: current_token_request::Type::Verified as i32,
        })
        .await
        .wrap_err("failed to get current token")?
        .into_inner();
    assert_response_valid(curr.header)?;
    let username = curr
        .token
        .ok_or_else(|| eyre::eyre!("Failed to get current token"))?
        .preferred_username;

    tracing::debug!("Fetching AWS Credentials...");
    let aws_creds = sts
        .assume_role_with_web_identity()
        .role_arn(opts.role_arn)
        .role_session_name(username)
        .web_identity_token(res.access_token)
        .send()
        .await
        .wrap_err("failed to assume AWS role")?;

    if let Some(c) = aws_creds.credentials {
        let cached = AWSCredentialOutput {
            version: 1,
            access_key_id: c.access_key_id,
            secret_access_key: c.secret_access_key,
            session_token: c.session_token,
            expiration: Timestamp {
                seconds: c.expiration.secs(),
                nanos: c.expiration.subsec_nanos() as i32,
            },
        };
        cc.set(cached.clone()).await.wrap_err("failed to cache AWS credentials")?;
        return Ok(cached);
    }
    bail!("No credentials received")
}
